File: app/src/main/java/org/ole/planet/myplanet/ui/team/AdapterTeam.java
package org.ole.planet.myplanet.ui.team;

import android.content.Context;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ListView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.recyclerview.widget.RecyclerView;

import org.ole.planet.myplanet.R;
import org.ole.planet.myplanet.databinding.ItemTeamGridBinding;
import org.ole.planet.myplanet.databinding.LayoutUserListBinding;
import org.ole.planet.myplanet.model.RealmMyTeam;
import org.ole.planet.myplanet.model.RealmUserModel;
import org.ole.planet.myplanet.utilities.Utilities;

import java.util.List;

import io.realm.Realm;

public class AdapterTeam extends RecyclerView.Adapter<AdapterTeam.ViewHolderTeam> {
    private ItemTeamGridBinding itemTeamGridBinding;
    private Context context;
    private List<RealmMyTeam> list;
    private Realm mRealm;
    private OnUserSelectedListener listener;
    private List<RealmUserModel> users;
    private OnTeamSelectedListener teamSelectedListener;

    public interface OnTeamSelectedListener {
        void onSelectedTeam(RealmMyTeam team);
    }

    public void setTeamSelectedListener(OnTeamSelectedListener teamSelectedListener) {
        this.teamSelectedListener = teamSelectedListener;
        Utilities.log("Team selected listener " + (teamSelectedListener == null));
    }

    public AdapterTeam(Context context, List<RealmMyTeam> list, Realm mRealm) {
        this.context = context;
        this.list = list;
        this.mRealm = mRealm;

        if (context instanceof OnUserSelectedListener) listener = (OnUserSelectedListener) context;
    }

    @NonNull
    @Override
    public ViewHolderTeam onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        itemTeamGridBinding = ItemTeamGridBinding.inflate(LayoutInflater.from(context), parent, false);
        return new ViewHolderTeam(itemTeamGridBinding);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolderTeam holder, int position) {
        itemTeamGridBinding.title.setText(list.get(position).getName());
        holder.itemView.setOnClickListener(view -> {
            if (this.teamSelectedListener != null)
                this.teamSelectedListener.onSelectedTeam(list.get(position));
            else showUserList(list.get(position));
        });
    }

    private void showUserList(RealmMyTeam realmMyTeam) {
        LayoutUserListBinding layoutUserListBinding = LayoutUserListBinding.inflate(LayoutInflater.from(context));
        users = RealmMyTeam.getUsers(realmMyTeam.get_id(), mRealm, "");
        setListAdapter(layoutUserListBinding.listUser, users);
        layoutUserListBinding.etSearch.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                users = RealmMyTeam.filterUsers(realmMyTeam.get_id(), charSequence.toString(), mRealm);
                setListAdapter(layoutUserListBinding.listUser, users);
            }

            @Override
            public void afterTextChanged(Editable editable) {
            }
        });
        layoutUserListBinding.listUser.setOnItemClickListener((adapterView, view1, i, l) -> {
            if (listener != null) listener.onSelectedUser(users.get(i));
        });
        new AlertDialog.Builder(context).setTitle(R.string.select_user_to_login).setView(layoutUserListBinding.getRoot()).setNegativeButton(R.string.dismiss, null).show();
    }

    private void setListAdapter(ListView lv, List<RealmUserModel> users) {
        ArrayAdapter<RealmUserModel> adapter = new ArrayAdapter<>(context, android.R.layout.simple_list_item_1, users);
        lv.setAdapter(adapter);
    }

    @Override
    public int getItemCount() {
        return list.size();
    }

    public interface OnUserSelectedListener {
        void onSelectedUser(RealmUserModel userModel);
    }

    static class ViewHolderTeam extends RecyclerView.ViewHolder {
        ItemTeamGridBinding itemTeamGridBinding;

        public ViewHolderTeam(ItemTeamGridBinding itemTeamGridBinding) {
            super(itemTeamGridBinding.getRoot());
            this.itemTeamGridBinding = itemTeamGridBinding;
        }
    }
}

File: app/src/main/java/org/ole/planet/myplanet/ui/team/AdapterTeam.kt
package org.ole.planet.myplanet.ui.team

import android.content.Context
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.ListView
import androidx.appcompat.app.AlertDialog
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import io.realm.Realm
import org.ole.planet.myplanet.R
import org.ole.planet.myplanet.databinding.ItemTeamGridBinding
import org.ole.planet.myplanet.databinding.LayoutUserListBinding
import org.ole.planet.myplanet.model.RealmMyTeam
import org.ole.planet.myplanet.model.RealmUserModel
import org.ole.planet.myplanet.utilities.DiffUtils

class AdapterTeam(
    private val context: Context,
    teams: List<RealmMyTeam>,
    private val mRealm: Realm
) : ListAdapter<RealmMyTeam, AdapterTeam.ViewHolderTeam>(
    DiffUtils.itemCallback(
        areItemsTheSame = { oldItem, newItem -> oldItem._id == newItem._id },
        areContentsTheSame = { oldItem, newItem -> oldItem.name == newItem.name }
    )
) {
    private var teamSelectedListener: OnTeamSelectedListener? = null
    private var listener: OnUserSelectedListener? = context as? OnUserSelectedListener
    private var users: List<RealmUserModel> = emptyList()

    init {
        submitList(teams)
    }

    interface OnTeamSelectedListener {
        fun onSelectedTeam(team: RealmMyTeam)
    }

    fun setTeamSelectedListener(teamSelectedListener: OnTeamSelectedListener?) {
        this.teamSelectedListener = teamSelectedListener
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolderTeam {
        val itemTeamGridBinding = ItemTeamGridBinding.inflate(LayoutInflater.from(context), parent, false)
        return ViewHolderTeam(itemTeamGridBinding)
    }

    override fun onBindViewHolder(holder: ViewHolderTeam, position: Int) {
        val team = getItem(position)
        holder.itemTeamGridBinding.title.text = team.name
        holder.itemView.setOnClickListener {
            teamSelectedListener?.onSelectedTeam(team) ?: showUserList(team)
        }
    }

    private fun showUserList(realmMyTeam: RealmMyTeam) {
        val layoutUserListBinding = LayoutUserListBinding.inflate(LayoutInflater.from(context))
        users = RealmMyTeam.getUsers(realmMyTeam._id, mRealm, "")
        setListAdapter(layoutUserListBinding.listUser, users)
        layoutUserListBinding.etSearch.addTextChangedListener(object : android.text.TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                users = RealmMyTeam.filterUsers(realmMyTeam._id, "$s", mRealm)
                setListAdapter(layoutUserListBinding.listUser, users)
            }
            override fun afterTextChanged(s: android.text.Editable) {}
        })
        layoutUserListBinding.listUser.setOnItemClickListener { _, _, position, _ ->
            listener?.onSelectedUser(users[position])
        }
        AlertDialog.Builder(context).setTitle(R.string.select_user_to_login)
            .setView(layoutUserListBinding.root)
            .setNegativeButton(R.string.dismiss, null)
            .show()
    }

    private fun setListAdapter(lv: ListView, users: List<RealmUserModel>) {
        val adapter = ArrayAdapter(context, android.R.layout.simple_list_item_1, users)
        lv.adapter = adapter
    }

    fun updateTeams(newTeams: List<RealmMyTeam>) {
        submitList(newTeams)
    }

    interface OnUserSelectedListener {
        fun onSelectedUser(userModel: RealmUserModel)
    }

    class ViewHolderTeam(val itemTeamGridBinding: ItemTeamGridBinding) :
        RecyclerView.ViewHolder(itemTeamGridBinding.root)
}
